### DISCUSSION TITLE
A hard lesson learned about icons in building iOS app recently against AIR SDK 33.1.1.544

### DISCUSSION THREAD

**Discussion by SUSDEV** [airsdk/Adobe-Runtime-Support#1046] [General]
Hi all, the past week I have been building an update to my iOS app in preparation for Harman releasing an update for iOS 15. I spent many many hours on a very odd error involving icons, so I thought I'd jot down some notes in case someone else comes across the issue.

TL;DR - use [this guide](https://airnativeextensions.github.io/tutorials/ios-icons-assets-car) to publish your icons for Assets.car using [appuploader tool](http://applicationloader.net/appuploader/icontool.php), not others you may find online. _Lesson_: If you get the errors I describe and can't find the cause, try regenerating the icons with another tool (there seem to be many online).

Now settle in for the full story...

The last update I pushed was with AIR SDK 33.1.1.216 I think, about 1 year ago, off my Catalina machine. This was before Apple's deadline of including `Assets.car` and launch storyboard (June 30, 2020). Also, I bought an M1 Mac-Mini with Big Sur recently, so there were a number of new factors to consider in getting my app updated.

- I updated my M1 to the latest version of Big Sur (c. Aug 4, 2021) and Xcode (12.5.1)
- - Oddly enough, on my M1 when I went to Keychain Assistant to create my CSR, it never showed me the window of certificates. 30 minutes of searching helped me find a console command I issued to show the list of items: `xattr -d com.apple.FinderInfo ~/Library` after which the list of certificates and such showed up in Keychain Assistant
- I created new development and distribution certificates and provisioning profiles in the Apple developer portal. My iPads are a bit old, with my newest one being an iPad Pro 2019. I found it curious that Apple has changed the UDID from a 40 character hex string to a 25 character (I think) string. Of course I verified my test device UDID was in the provisioning profile

- The task I have been avoiding was to create `Assets.car` and storyboard. G-d bless the Distriqt folk(s) for making [this guide](https://airnativeextensions.github.io/tutorials/ios-icons-assets-car), it literally was a life saver. I ended up using Xcode to generate Assets.car and the storyboard, and even though my current Xcode is slightly different, it was pretty straightforward -- except for the icons part.
- As [the guide](https://airnativeextensions.github.io/tutorials/ios-icons-assets-car) describes, there are online tools for making Assets.car. The guide suggested using [applicationloader.net](http://applicationloader.net/appuploader/icontool.php), which ultimately was the hero. For some reason, I had used initially a different tool online to generate the set of icons, so definitely use that link instead of whatever else you might find.
- The guide also has a link to making the storyboard, which was easy with Xcode. I am attaching PDF files of these guides in case the page cannot be loaded for some reason
[Making Storyboard for iOS - by Distriqt.pdf](https://github.com/airsdk/Adobe-Runtime-Support/files/6935054/Making.Storyboard.for.iOS.-.by.Distriqt.pdf)
[Making Assets-car for iOS - by Distriqt.pdf](https://github.com/airsdk/Adobe-Runtime-Support/files/6935055/Making.Assets-car.for.iOS.-.by.Distriqt.pdf)
- I had trouble building a development version of my app. Whenever I tried to drop it onto my iPad (connected my iPad to my Mac), it told me it could not verify the integrity of the app, so it would not load.
- Therefore, to get things rolling, I made an ad hoc distribution using the first set of icons (the troubled ones). That worked perfectly fine, so I had a false sense of security.
- I built a production/distribution version of my app (one with a provisioning profile for the app store) with Assets.car built from the first set of icons I had generated, but I thought I was set since it worked for the ad hoc limited distribution release.
- I then went into the Apple developer portal and created a new version entry of my app with the version # from the app I just built
- on the console, I ran this script I made (put in your username/password) to upload for TestFlight:

> #!/bin/bash
> 
> set -euo pipefail
> 
> xcrun altool --upload-app --type ios --file "/Users/.../ios/SimsUShareMobileIOS.ipa" --username "<your apple id email>" --password "<your app password>"

- Frankly, I forget where I had set the password for this, so maybe someone can comment where you get the password. I think it was an app-specific password I made with TestFlight but I don't recall
- Then the fun started. I got errors about icons not being included. I made another small script to validate the upload, instead of trying repeatedly to use the `upload-app` option of altool:

> #!/bin/bash
> 
> set -euo pipefail
> 
> xcrun altool --validate-app --type ios --file "/Users/.../ios/SimsUShareMobileIOS.ipa" --username "<your app login>" --password "<your app password>"

- Being a little paranoid, I also had an experience the last time out in which the upload-app failed because of not finding altool. I issued the following to reset it: `sudo xcode-select -r`
- Here are the errors I got after using altool:
`Mac-mini:ios jonathankaye$ ./validate.sh 
*** Error: Unable to validate archive '/Users/.../ios/SimsUShareMobileIOS.ipa'.
*** Error: Missing required icon file. The bundle does not contain an app icon for iPad of exactly '76x76' pixels, in .png format for iOS versions >= 7.0. To support older operating systems, the icon may be required in the bundle outside of an asset catalog. Make sure the Info.plist file includes appropriate entries referencing the file. See https://developer.apple.com/documentation/bundleresources/information_property_list/user_interface App Store operation failed. (1091)
*** Error: Missing required icon file. The bundle does not contain an app icon for iPad of exactly '152x152' pixels, in .png format for iOS versions >= 7.0. To support older operating systems, the icon may be required in the bundle outside of an asset catalog. Make sure the Info.plist file includes appropriate entries referencing the file. See https://developer.apple.com/documentation/bundleresources/information_property_list/user_interface App Store operation failed. (1091)
*** Error: Missing App Icon. An app icon measuring 1024 by 1024 pixels in PNG format must be included in the Asset Catalog of apps built for iOS, iPadOS, or watchOS. Without this icon, apps cannot be submitted for review. For details, see https://developer.apple.com/ios/human-interface-guidelines/icons-and-images/app-icon/. App Store operation failed. (1091)`

- I looked online for days and tried all sorts of 'solutions' but nothing worked (one full build cycle for me took about 15 minutes, so it was somewhat painful), things like modifying info.plist, and all sorts of craziness. I was absolutely certain I had packaged the 1024x1024 icon in Assets.car and I had the other icons outside As a last resort, I tried replacing the original icons I had put in with the set I made from appuploader. Voila, the errors disappeared on validate as well as submit!
- I don't really care why the first set of icons did not work though they looked exactly correct, but this was certainly a Hail Mary operation.
- I got everything submitted to TestFlight and tried it on my iPad Pro, and it is working. Frankly, I don't see the launch image screen I made and it seems to flash the old screen (pre-storyboard), but I will wait to see what Apple says eventually when I submit it for review -- and keep my fingers crossed.

Well, maybe others will be inspired to write about their journey in publishing apps and obstacles they overcame. The Internet never forgets!

-jonathan

**Comment by PippoApps** [comment]
Cheers for the detailed walkthrough mate.
Saved me a lot of headache.
